### ISSUE TITLE
29-bit CAN IDs incorrectly interpreted as IDE = 0 if they appear as 11-bit

### ISSUE THREAD

**Issue by MatinF** [danielhrisca/asammdf#905]
Tested in latest dev version.

If an ID appears as if it's an 11-bit ID, e.g. `000000C9`, it is apparently treated as 11-bit in the DBC decoding logic. My guess is that the ID is marked this way based on the actual ID string (removing prefix 0's), rather than using the IDE.

This can be hard to troubleshoot because the `IDE` field is no longer part of the CAN bus trace view (recommend including it again).

I can send two files that reflect the issue if needed. The first file has the ID `0C9` as an actual 11-bit ID with IDE = 0. The second file has the ID `000000C9` with IDE = 1, but it is still decoded as if it was IDE = 0.

**Comment by danielhrisca** [comment]
@MatinF please send the files

**Comment by MatinF** [comment]
You can also see it via the CAN bus trace more easily via the latest dev version :-)

![image](https://github.com/danielhrisca/asammdf/assets/26184427/baab4c82-d41b-4459-b18b-e0df39eefad9)

**Comment by danielhrisca** [comment]
According to the MDF bus logging specifications, the ID filed should contain the full extended value (unmasked). The DBC file needs to be fixed as well since all the messages are defined as standard IDs , and will not be compatible with the J1939 checks done in canmatrix

![image](https://github.com/danielhrisca/asammdf/assets/20952040/cd998b68-c2c9-4c15-8867-6fe7d0d76564)

**Comment by MatinF** [comment]
Hi Daniel,

1) The DBC file had a syntax error, this is fixed here:
[fca_left_v2.zip](https://github.com/danielhrisca/asammdf/files/13077613/fca_left_v2.zip)

2) I am not sure I understand the part regarding J1939 - the DBC file is a DBC file related to standard 11-bit CAN IDs, it's not supposed to be a J1939 DBC or log file. Rather, the "problem" in this case is that a 29-bit CAN ID, `0x000000C9`, is correctly marked with `IDE` 1 (meaning it's extended 29-bit) is treated in the GUI as if it was an 11-bit ID `0xC9` with `IDE` 0. 

3) Our understanding is that the 'legacy' method for registering a 29-bit ID in the MDF file format was to use 32 bits, which is what many Vector tools do. However, our understanding is that the standard allows for simply storing just the 29-bit ID and using the IDE field alone to show whether an IDE is extended or not - without having to store the 32 bit version of the 29 bit ID in the MF4 file. Vector tools previously would only show an ID as extended if the 32 bit ID method was used in an MDF, but they now also show them as extended if the 29 bit method is used as in our MDF files. I can check with the team for details on this if you believe the standard is to be interpreted differently. It's worth noting that all our MF4 files use the 29 bit method and everything is correctly processed in asammdf, meaning that all 29-bit (incl. J1939 PGN style) decoding works without requiring the use of 32 bits in our log files. The only problem seems to be the case where an ID is in fact extended (based on the IDE), but where the ID takes up less than 29 bit as in the case as is the case for the ID in question.


I believe the correct solution here is that the asammdf code is adjusted to allow the IDE field to dictate whether an ID is extended or not - regardless of the bit length of the ID.

**Comment by danielhrisca** [comment]
I've modified the DBC to contain the message ID 0xC9 both as standard and extended addressing modes. 

[fca_left_v3.zip](https://github.com/danielhrisca/asammdf/files/13112111/fca_left_v3.zip)

Please try the development branch code and check if the output is correct:

![image](https://github.com/danielhrisca/asammdf/assets/20952040/7b41ebdd-30d3-4ad0-a435-6e55126473a0)

**Comment by MatinF** [comment]
Hi Daniel,

Just to clarify: The DBC in fca_left_v2.zip is correct - i.e. it is deliberately supposed to _not_ decode any data in the 2nd file. The purpose is to showcase that the CAN ID `0x000000C9` should not be matched by the v2 DBC because the IDE marks the ID `0x000000C9` as extended in the 2nd log file. 

To make it a bit more clear:

1) The log file in `2023-08-30/` contains only an 11-bit version of the ID `0xC9` that should be matched by the `fca_left_v2.dbc`
2) The log file in `2023-09-04/` contains only a 29-bit version of the ID `0x000000C9` that should _not_ be matched by the `fca_left_v2.dbc` (but which is currently being incorrectly matched in the latest dev version)

In other words, the issue is not with the DBC file provided - but rather with the fact that the current asammdf code does not rely on the `IDE` field to determine if an ID is extended or not when it comes to DBC decoding.

I hope this clarifies, else let me know.

**Comment by danielhrisca** [comment]
> 1. The log file in `2023-08-30/` contains only an 11-bit version of the ID `0xC9` that should be matched by the `fca_left_v2.dbc`

The file contains 0xC9 with both 11bit and 29bit addressing

**Comment by MatinF** [comment]
Hi again,

You are right that the first file is a mix, sorry for missing that.

However, the second file in `2023-09-04/` only contains `IDE = 0x01` from what I can tell:
![image](https://github.com/danielhrisca/asammdf/assets/26184427/6719d2e9-e6bb-4099-8c24-8787c01cbd87)

This is the file that I've mainly focused on. 

The problem can be observed in both files, however.

If you look in the first file, the expected behavior is that the last matched observation by the DBC should be at time 602.8466 as per below:
![image](https://github.com/danielhrisca/asammdf/assets/26184427/c4fed1a7-e1a9-410d-a215-94c275ffe7bc)

However, if you decode the first file with the DBC you'll see that the signal continues beyond 602.8466 as per below:
![image](https://github.com/danielhrisca/asammdf/assets/26184427/63f1e273-1858-49f4-9a6e-c450a81d558c)

**Comment by danielhrisca** [comment]
Please grab the development code again, you are missing some commits.
With the new code I get this using the v2.dbc

![image](https://github.com/danielhrisca/asammdf/assets/20952040/39248d3f-796b-4656-ac63-45c267668f0d)

**Comment by MatinF** [comment]
Great, it looks like it is resolved in the latest dev version! Thanks Daniel
